server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -101,17 +107,44 @@ public ShardId getShardId() {
 
     public boolean transferSnapshot(TransferSnapshot transferSnapshot, TranslogTransferListener translogTransferListener)
         throws IOException {
-        List<Exception> exceptionList = new ArrayList<>(transferSnapshot.getTranslogTransferMetadata().getCount());
-        Set<TransferFileSnapshot> toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
         long metadataBytesToUpload;
         long metadataUploadStartTime;
         long uploadStartTime;
         long prevUploadBytesSucceeded = remoteTranslogTransferTracker.getUploadBytesSucceeded();
         long prevUploadTimeInMillis = remoteTranslogTransferTracker.getTotalUploadTimeInMillis();
 
         try {
-            toUpload.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getTranslogFileSnapshots()));
-            toUpload.addAll(fileTransferTracker.exclusionFilter((transferSnapshot.getCheckpointFileSnapshots())));
+            boolean isObjectMetadataUploadSupportedInBlobStore = transferService.isObjectMetadataUploadSupported();
+
+            List<Exception> exceptionList = new ArrayList<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+            Set<TransferFileSnapshot> toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+
+            // if the transferService of enabled blob store supports uploading object metadata, we don't need to transfer checkpoint file
+            // snapshots separately. We will transfer checkpoint data as metadata to translog file upload object.
+            if (isObjectMetadataUploadSupportedInBlobStore) {
+                Set<Tuple<TransferFileSnapshot, TransferFileSnapshot>> tlogAndCkpFilesTupleSet = transferSnapshot
+                    .getTranslogAndCheckpointFileSnapshotTupleSet();
+
+                for (Tuple<TransferFileSnapshot, TransferFileSnapshot> tuple : tlogAndCkpFilesTupleSet) {
+                    TransferFileSnapshot translogSnapshot = tuple.v1();
+                    TransferFileSnapshot checkpointSnapshot = tuple.v2();
+                    if (!fileTransferTracker.isAlreadyUploaded(translogSnapshot)) {
+                        Map<String, String> metadata = createCheckpointDataAsObjectMetadata(checkpointSnapshot);
+                        translogSnapshot.setTransferFileSnapshotMetadata(metadata);
+                        translogSnapshot.setTransferFileSnapshotMetadataFileName(checkpointSnapshot.getName());
+
+                        // close the checkpoint fileSnapshot now
+                        checkpointSnapshot.close();
```

Comment:
Why is this required?

Reply:
The reason we need to close the CheckpointFileSnapshot is to ensure that the underlying FileChannel associated with the snapshot is properly closed and released.